server/handlers/project_command_output_handler.go

```diff
@@ -0,0 +1,134 @@
+	//     return err
+	//   }
+	// })
+	Receive(projectInfo string, callback func(msg string) error) error
+
+	// Is basically the same as Listen function from logstreaming controller.
+	Handle()
+}
+
+func NewProjectCommandOutputHandler(projectCmdOutput chan *models.ProjectCmdOutputLine, logger logging.SimpleLogging) ProjectCommandOutputHandler {
+	return &DefaultProjectCommandOutputHandler{
+		ProjectCmdOutput:  projectCmdOutput,
+		logger:            logger,
+		controllerBuffers: map[string]map[chan string]bool{},
+	}
+}
+
+func (p *DefaultProjectCommandOutputHandler) Send(ctx models.ProjectCommandContext, msg string) {
+	p.ProjectCmdOutput <- &models.ProjectCmdOutputLine{
+		ProjectInfo: ctx.PullInfo(),
+		Line:        msg,
+	}
+}
+
+func (p *DefaultProjectCommandOutputHandler) Receive(projectInfo string, callback func(msg string) error) error {
+	ch := p.addChan(projectInfo)
+	defer p.removeChan(projectInfo, ch)
+
+	for msg := range ch {
+		if err := callback(msg); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *DefaultProjectCommandOutputHandler) Handle() {
+	for msg := range p.ProjectCmdOutput {
+		p.logger.Info("Recieving message %s", msg.Line)
+		if msg.ClearBuffBefore {
+			p.clearLogLines(msg.ProjectInfo)
+		}
+		p.writeLogLine(msg.ProjectInfo, msg.Line)
+		if msg.ClearBuffAfter {
+			p.clearLogLines(msg.ProjectInfo)
+		}
+	}
+}
+
+func (p *DefaultProjectCommandOutputHandler) clearLogLines(pull string) {
+	p.controllerBufferLock.Lock()
+	delete(p.projectOutputBuffers, pull)
+	p.controllerBufferLock.Unlock()
+}
+
+func (p *DefaultProjectCommandOutputHandler) addChan(pull string) chan string {
+	ch := make(chan string, 1000)
+	p.controllerBufferLock.Lock()
+	for _, line := range p.projectOutputBuffers[pull] {
+		ch <- line
+	}
+	if p.controllerBuffers == nil {
+		p.controllerBuffers = map[string]map[chan string]bool{}
+	}
+	if p.controllerBuffers[pull] == nil {
+		p.controllerBuffers[pull] = map[chan string]bool{}
+	}
+	p.controllerBuffers[pull][ch] = true
+	p.controllerBufferLock.Unlock()
+	return ch
+}
+
+//Add log line to buffer and send to all current channels
+func (p *DefaultProjectCommandOutputHandler) writeLogLine(pull string, line string) {
+	p.controllerBufferLock.Lock()
+	if p.projectOutputBuffers == nil {
+		p.projectOutputBuffers = map[string][]string{}
+	}
+	p.logger.Info("Project info: %s, content: %s", pull, line)
+
+	for ch := range p.controllerBuffers[pull] {
+		select {
+		case ch <- line:
+		default:
+			delete(p.controllerBuffers[pull], ch)
+		}
+	}
+	if p.projectOutputBuffers[pull] == nil {
+		p.projectOutputBuffers[pull] = []string{}
+	}
+	p.projectOutputBuffers[pull] = append(p.projectOutputBuffers[pull], line)
+	p.controllerBufferLock.Unlock()
+}
+
+//Remove channel, so client no longer receives Terraform output
+func (p *DefaultProjectCommandOutputHandler) removeChan(pull string, ch chan string) {
+	p.controllerBufferLock.Lock()
+	delete(p.controllerBuffers[pull], ch)
```

Comment:
We should also close the channel here. 
```suggestion
	delete(p.controllerBuffers[pull], ch)
	close(ch)
```